walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
+            Math.max(optionalBlockExtent, mandatoryBlockExtent), // Only use overlap between optional/mandatory availability.
+            when.start(), false)
+        );
+      }
+      
+
+      if (containsMandatoryAttendees && mandatoryWindowLongEnough) {
+        timeSlots.add(
+          /* Add non-inclusive window because it can't overlap with the start of this event. */
+          TimeRange.fromStartEnd(mandatoryBlockExtent, when.start(), false)
+        );
+      }
+
+      // If this window contains members from a certain class, update their respective block extent.
+      if (containsOptionalAttendees) { // We group all optional attendees as one lump for now.
+        optionalBlockExtent = Math.max(optionalBlockExtent, when.end());
+      }
+      if (containsMandatoryAttendees) {
+        mandatoryBlockExtent = Math.max(mandatoryBlockExtent, when.end());
+      }
+    }
+
+    // At this point there are no more events, so remaining might be a contiguous potential window.
+    boolean potentialMandatoryWindowExists = windowIsLongEnough(mandatoryBlockExtent, TimeRange.END_OF_DAY, request);
+    // It is only acceptable for the extent to still be 0 IF there are no events that could move that extent.
+    boolean remainingMandatoryExtentAcceptable = mandatoryBlockExtent > 0 || hasMandatoryEvents == false;
+
+    // Same as above, for optional.
+    boolean potentialOptionalWindowExists = windowIsLongEnough(optionalBlockExtent, TimeRange.END_OF_DAY, request);
+    boolean remainingOptionalExtentAcceptable = optionalBlockExtent > 0 || hasOptionalEvents == false;
+
+    // This could be a mandatory-window.
+    if (potentialMandatoryWindowExists && remainingMandatoryExtentAcceptable) {
+      // Per test FindMeetingQueryTest.optionsForNoAttendees, having no attendees should still return availability.
+      boolean noAttendeesToConsider = !considerMandatoryAttendees && !considerOptionalAttendees;
+      if (considerMandatoryAttendees || noAttendeesToConsider) {
+        timeSlots.add(
+          TimeRange.fromStartEnd(mandatoryBlockExtent, TimeRange.END_OF_DAY, true)
+        );
+      }
+      // As above, we can only consider the optional window IF the mandatory window also exists.
+      if (considerOptionalAttendees && potentialOptionalWindowExists && remainingOptionalExtentAcceptable) {
+        timeSlotsWithOptional.add(
+          TimeRange.fromStartEnd(
+            Math.max(mandatoryBlockExtent, optionalBlockExtent), // Only use the overlap between optional/mandatory availability.
+            TimeRange.END_OF_DAY, true
+          )
+        );
+      }
+    }
+
+    // If possible, return the windows that include all optional. Otherwise, just return mandatory.
+    return timeSlotsWithOptional.size() > 0 ? timeSlotsWithOptional : timeSlots;
+  }
+
+  /**
+   * Ensure a potential window is long enough to fit meet request.
+   * @param from     start of window in minutes.
+   * @param end      end of window in minutes.
```

Comment:
When talking about timeranges, try to be explicit about whether times are inclusive or exclusive (is 10 to 20 a 9, 10, or 11 minute interval?)